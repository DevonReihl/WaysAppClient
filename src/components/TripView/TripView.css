.trip {
  width: 100%;
  border-radius: 1px;
  text-align: center;
  display: inline-flex;
  flex-direction: column;
  color: #494949;
  word-wrap: break-word;
}

.trip-stop {
  word-wrap: break-word;
  position: relative;
}

.trip-right {
  width: 100%;
}

.trip-left {
  width: 100%;
  max-height: 200px;
  border-radius: 6px;
  overflow: hidden;
}

.road-img {
  width: 70px;
  margin-top: 8px;
}

.belowMap {
  margin-top: 40px;
}

.trip-stop-wrapper {
  counter-increment: index;
  width: 98%;
  word-wrap: break-word;
  position: relative;
  z-index: 1;
}

.edit-stop {
  color: black;
  background-color: white;
  padding: 3px;
  display: inline-block;
  line-height: 1.6em;
  outline: none;
  position: relative;
  font-size: 1.3em;
  border: 5px solid rgb(46, 140, 216);
  border-radius: 15px 15px 15px 15px;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 98%;
}

.trip-name {
  font-size: 52px;
  text-decoration: none;
  color: rgb(46, 140, 216);
  text-shadow: 2px 1px 2px rgb(0, 0, 0);
  font-weight: bold;
  text-align: center;
  font-family: 'Fredoka One', Arial, Helvetica, sans-serif;
  margin: 0px;
  word-wrap: break-word;
}

.trip-info > p {
  color: rgb(141, 141, 141);
  font-family: 'Ubuntu', sans-serif;
  font-size: 24px;
  font-weight: bold;
}

.trip-info > p > span {
  font-size: 28px;
  color: #6b6b6b;
  text-align: center;
  font-family: Helvetica, sans-serif;
  font-weight: bold;
  letter-spacing: 1px;
  margin: 15px;
}

.trip-header > h2 {
  font-size: 40px;
  margin-top: 18px;
  font-family: 'Ubuntu', sans-serif;
}

.trip-header > div > input {
  font-size: 40px;
  border: none;
  text-align: center;
  margin: 0;
}

.trip-header-creator > h2 {
  font-size: 40px;
  margin-top: 3px;
  font-family: 'Ubuntu', sans-serif;
}

.trip-div {
  color: black;
  background-color: white;
  padding: 0px 2px;
  display: inline-block;
  margin: 19px 0px !important;
  line-height: 1.6em;
  outline: none;
  position: relative;
  font-size: 1.3em;
  border: 5px solid rgb(223, 223, 223);
  border-radius: 15px 15px 15px 15px;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 98%;
}

.trip-div:before {
  content: '';
  position: absolute;
  left: -5px;
  top: -5px;
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 35px 35px 0 0;
  border-color: rgb(230, 31, 130) transparent transparent transparent;
  z-index: 1;
}

.trip-div:hover,
.trip-div:hover {
  border-color: rgb(230, 31, 130);
  background-color: rgb(255, 255, 255);
  transform: scale(1.1);
}

.tripViewButton {
  box-shadow: inset 0px 1px 0px 0px #ffffff;
  background: linear-gradient(to bottom, #ffffff 5%, #f6f6f6 100%);
  background-color: #ffffff;
  border-radius: 6px;
  border: 1px solid #dcdcdc;
  display: inline-block;
  cursor: pointer;
  color: #5a5a5a;
  font-family: 'Ubuntu', sans-serif;
  font-size: 15px;
  font-weight: bold;
  padding: 2px 3px;
  text-decoration: none;
  text-shadow: 0px 1px 0px #ffffff;
}
.tripViewButton:hover {
  background: linear-gradient(to bottom, #f6f6f6 5%, #ffffff 100%);
  background-color: #f6f6f6;
}
.tripViewButton:active {
  position: relative;
  top: 1px;
}

.tripView-button-wrapper {
  text-align: right;
}

.trip-div figcaption:before {
  -webkit-transform: translateX(-50%);
  transform: translateX(-50%);
  background-color: #fff;
  border-radius: 50%;
  box-shadow: 0 0 10px rgba(117, 117, 117, 0.25);
  color: rgb(47, 47, 47);
  content: counter(index);
  font-family: 'Fredoka One', Arial, Helvetica, sans-serif;
  font-size: 17px;
  font-weight: bold;
  right: -4%;
  line-height: 45px;
  position: absolute;
  bottom: -18px;
  width: 45px;
}

#EditStopForm > div > input,
#EditStopForm > input {
  border: none;
  font-size: 26px;
  text-align: center;
}

.trip-category {
  text-transform: capitalize;
}

.addStopForm {
  padding: 20px;
  width: 100%;
  margin: auto;
  margin-top: 6px;
  border: 4px solid rgb(238, 238, 238);
  border-radius: 6px;
  position: relative;
  margin-bottom: 50px;
}

.addStopForm > label {
  font-size: 20px;
  font-weight: bold;
}

.addStopForm > input {
  margin-bottom: 12px;
}

.addStopForm > input::-webkit-input-placeholder {
  color: rgb(184, 184, 184);
}

.addStopForm > h2 {
  font-family: 'Fredoka One', Arial, Helvetica, sans-serif;
  font-size: 44px;
  margin: 0;
  color: rgb(233, 159, 64);
  text-shadow: 2px 1px 2px rgb(0, 0, 0);
}

.edit-trip > form > input {
  border: none;
  text-align: center;
  font-size: 26px;
  font-weight: bold;
}

.flikr-img {
  width: 300px;
  border-radius: 6px;
  padding-top: 5px;
}

.trip-flex {
  display: inline-flex;
  padding: 5px;
  flex-wrap: wrap;
  flex-direction: column;
  align-items: center;
}

.stop-description {
  text-align: center;
  width: 98%;
  font-size: 18px;
}

.buffer-img-wrapper {
  position: relative;
}

.fade-out-screen {
  background-color: rgb(255, 255, 255);
  z-index: 49;
  width: 1200px;
  height: 1200px;
  position: absolute;
  opacity: 0.7;
}

.buffer-img {
  position: absolute;
  z-index: 50;
  transform: translate(-50%, 55%);
  margin-left: 35px;
}

.like-btn {
  border: none;
}

img.heart {
  vertical-align: middle;
  width: 4em;
  transition: 0.5s ease-in-out;
  cursor: pointer;
}

#Map {
  flex: 1;
  margin-top: 0px;
  margin-bottom: 40px;
  margin: auto;
  border: 1px rgb(245, 245, 245) solid;
  min-height: 400px;
}

.trip-info {
  flex: 1;
}

.tripView-upperSection {
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
  width: 100%;
  padding: 0px;
}

.tripView-upperSection {
  flex-direction: column;
}

.trip-rating-digits {
  height: 80px;
  width: 315px;
  color: #f5f5f5;
  text-shadow: 1px 2px 2px black;
  font-family: 'Fredoka One', Arial, Helvetica, sans-serif;
  font-size: 20px;
}

.addStopBtn:hover {
  cursor: pointer;
}

@media all and (min-width: 675px) {
  .addStopForm {
    width: 675px;
  }

  .addStopForm > input {
    margin-bottom: 12px;
  }

  .edit-trip > form > input {
    width: 675px;
  }

  .tripView-upperSection {
    flex-direction: row;
    padding: 40px;
    padding-right: 60px;
    padding-left: 60px;
  }

  .trip-info {
    margin-right: 10px;
  }
}

@media all and (max-width: 675px) {
  #Map {
    width: 100%;
  }
}

@media all and (max-width: 350px) {
  .TripCardDemo-wrapper > img {
    width: 30px;
    height: 30px;
  }

  .TripCard-bottom-info {
    width: 200px;
  }
}

@media all and (min-width: 675px) {
  .trip-stop-wrapper {
    width: 100%;
  }

  .trip-div,
  .edit-stop {
    width: 675px;
    margin: auto;
  }

  .trip-flex {
    width: 675px;
    flex-direction: row;
  }

  .trip-right {
    width: 330px;
  }

  .trip-left {
    width: 330px;
  }
}
