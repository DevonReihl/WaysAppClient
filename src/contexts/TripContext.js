import React from 'react';
import UserContext from './UserContext';

const TripContext = React.createContext({
  trips: [],
  currTripId: null,
  setCurrTripId: () => {},
  setTrips: () => {},
  verifyAuth: () => {},
  setLoading: () => {},
  returnUserID: () => {},
  setTripRating: () => {},
  loading: null,
});

export default TripContext;

export class TripProvider extends React.Component {
  state = {
    trips: [],
    currTripId: null,
    currPlace: {},
    stops: [],
    loading: false,
  };

  static contextType = UserContext;

  setTrips = (res) => {
    this.setState({
      trips: res,
    });
  };

  setTripRating = (trip_id) => {
    const index = this.state.trips.findIndex((trip) => {
      return trip.id === Number(trip_id);
    });
    let trips = [...this.state.trips];
    trips[index] = {
      ...trips[index],
      rating: Number(trips[index].rating) + 1,
    };
    this.setState({ trips });
  };

  setLoading = (boolean) => {
    this.setState({
      loading: boolean,
    });
  };

  setCurrTripId = (id) => {
    this.setState({
      currTripId: id,
    });
  };

  handleAddTrip = (trip) => {
    this.setState({
      trips: [...this.state.trips, trip],
    });
  };

  verifyAuth = (id) => {
    return id === this.context.user.id;
  };

  returnUserID = () => {
    return this.context.user.id;
  };

  render() {
    const value = {
      trips: this.state.trips,
      currTripId: this.state.currTripId,
      setTrips: this.setTrips,
      setCurrTripId: this.setCurrTripId,
      verifyAuth: this.verifyAuth,
      loading: this.state.loading,
      setLoading: this.setLoading,
      returnUserID: this.returnUserID,
      setTripRating: this.setTripRating,
    };

    return (
      <TripContext.Provider value={value}>
        {this.props.children}
      </TripContext.Provider>
    );
  }
}
