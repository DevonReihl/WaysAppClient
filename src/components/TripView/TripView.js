import React from 'react';
import Modal from 'react-modal';
import TripApiService from '../../services/trip-service';
import TripContext from '../../contexts/TripContext';
import TripViewNav from './TripViewNav/TripViewNav';
import RenderStop from './RenderStop/RenderStop';
import MapContainer from '../Map/Map';
import AddStopForm from './AddStopForm.js/AddStopForm';
import Helpers from '../../helpers/helpers';
import EditStopForm from './EditStopForm/EditStopForm';
import './TripView.css';
import images from '../../assets/images/images';
import RenderEditTrip from './RenderEditTrip/RenderEditTrip';

// Modal.setAppElement('#root');

export default class Trip extends React.Component {
  static contextType = TripContext;
  state = {
    stops: [],
    trip: [{ user_id: 0, short_description: 'Add a Stop!' }],
    currTripID: 0,
    tripDescription: '',
    toggleAddStop: false,
    stopEditingID: 0,
    toggleEditTrip: false,
    selections: [],
    error: null,
    userHasRated: false,
    isModalOpen: false,
  };

  componentDidMount() {
    const trip_id = this.props.match.params.trips_id;
    this.userHasRated();
    this.context.setLoading(true);
    TripApiService.getTrip(trip_id)
      .then((res) => {
        this.setState({ trip: res, currTripID: res.id });
      })
      .catch((res) => {
        this.setState({ error: res.error });
      })
      .finally(() => {
        this.context.setLoading(false);
      });
    this.context.setLoading(true);
    TripApiService.getStops(trip_id)
      .then((res) => {
        this.setState({ stops: res });
      })
      .catch((res) => {
        this.setState({ error: res.error });
      })
      .finally(() => {
        this.context.setLoading(false);
      });
  }

  toggleAddStop = () => {
    this.setState({ toggleAddStop: !this.state.toggleAddStop });
  };

  toggleEditStop = (stop_id, stop_categories) => {
    if (stop_id !== 0) {
      this.setState({
        stopEditingID: stop_id,
        selections: stop_categories.split(','),
      });
    } else {
      this.setState({
        stopEditingID: stop_id,
      });
    }
  };

  handleSelect = (e) => {
    let selection = e.target;
    selection = selection.innerHTML;
    const findSelect = this.state.selections.includes(selection);
    if (findSelect) {
      this.setState({
        selections: [
          ...this.state.selections.filter((select) => {
            return select !== selection;
          }),
        ],
      });
    } else {
      this.setState({ selections: [...this.state.selections, selection] });
    }
  };

  clearSelections = () => {
    this.setState({ selections: [] });
  };

  handleDeleteTrip = () => {
    this.context.setLoading(true);
    TripApiService.deleteTrip(this.state.trip[0].id)
      .then(() => {
        this.context.setTrips(
          this.context.trips.filter((trip) => trip.id !== this.state.trip[0].id)
        );
        this.props.history.push('/dashboard');
      })
      .catch((res) => {
        this.setState({ error: res.error });
      })
      .finally(() => {
        this.context.setLoading(false);
      });
  };

  handleEditTrip = () => {
    this.setState({ toggleEditTrip: true });
  };

  handleSubmitEditedTrip = (e, id) => {
    e.preventDefault();
    const { short_description, activities, days } = e.target;
    let trip = {
      short_description: short_description.value,
      activities: activities.value,
      days: days.value,
    };
    this.context.setLoading(true);
    TripApiService.patchTrip(trip, id)
      .then((res) => {
        let rating = this.state.trip[0].rating;
        res[0].rating = rating;
        this.setState({ trip: res });
      })
      .catch((res) => {
        this.setState({ error: res.error });
      })
      .finally(() => {
        this.context.setLoading(false);
        this.setState({ toggleEditTrip: false });
      });
  };

  handleSubmitEditStop = async (e, stop_id) => {
    e.preventDefault();
    if (this.state.selections.length === 0) {
      this.setState({ error: 'Must select atleast 1 Category' });
      return;
    }
    this.setState({
      error: null,
      stopEditingID: 0,
    });
    const { stop_name, description, city, state } = e.target;
    let tripId = this.props.match.params.trips_id;
    let stop = {
      trip_id: tripId,
      longitude: 0.0,
      latitude: 0.0,
      city: city.value,
      state: state.value,
      stop_name: stop_name.value,
      description: description.value,
      category: this.state.selections.join(', '),
    };
    this.context.setLoading(true);
    const res = await Helpers.flickrApi(stop_name.value, city.value);
    stop.img = Helpers.generateFlikrLink(res);
    TripApiService.patchStop(stop, stop_id)
      .then((res) => {
        const stops = this.state.stops.filter((stop) => stop.id !== res.id);
        this.setState({
          stops: [...stops, res],
          selections: [],
        });
      })
      .catch((res) => {
        this.setState({ error: res.error });
      })
      .finally(() => {
        this.context.setLoading(false);
      });
  };

  handleSubmitStop = async (e) => {
    e.preventDefault();
    if (this.state.selections.length === 0) {
      this.setState({ error: 'Must select atleast 1 Category' });
      return;
    }
    this.setState({ error: null });
    const { stop_name, description, city, state } = e.target;
    let tripId = this.props.match.params.trips_id;
    let stop = {
      trip_id: tripId,
      longitude: -32.77779,
      latitude: 46.888888,
      city: city.value,
      state: state.value,
      stop_name: stop_name.value,
      description: description.value,
      category: this.state.selections.join(', '),
    };
    this.context.setLoading(true);
    const res = await Helpers.flickrApi(stop_name.value, city.value);
    stop.img = Helpers.generateFlikrLink(res);
    TripApiService.postStop(stop)
      .then((res) => {
        this.setState({
          stops: [...this.state.stops, res],
          toggleAddStop: false,
          selections: [],
        });
      })
      .catch((res) => {
        this.setState({ error: res.error });
      })
      .finally(() => {
        this.context.setLoading(false);
      });
  };

  handleRating = () => {
    const trip_id = this.props.match.params.trips_id;
    const user_id = this.context.returnUserID();
    const rate = 1;
    const rating = { trip_id, user_id, rate };
    if (user_id !== undefined) {
      TripApiService.postRating(rating)
        .then(() => {
          this.setState({
            trip: [
              {
                ...this.state.trip[0],
                rating: Number(this.state.trip[0].rating) + 1,
              },
            ],
            userHasRated: true,
          });
          this.context.setTripRating(trip_id);
        })
        .catch((res) => {
          this.setState({ error: res.error });
        });
    }
  };

  toggleModal = (boolean) => {
    this.setState({ isModalOpen: boolean });
  };

  userHasRated = () => {
    const trip_id = this.props.match.params.trips_id;
    const user_id = this.context.returnUserID();
    this.context.setLoading(true);
    if (user_id !== undefined) {
      TripApiService.checkUserHasRated(trip_id)
        .then((res) => {
          for (let i = 0; i < res.length; i++) {
            if (res[i].user_id === user_id) {
              this.setState({ userHasRated: true });
              this.context.setLoading(false);
              return;
            }
          }
        })
        .catch((res) => {
          this.setState({ error: res.error });
        })
        .finally(() => {
          this.context.setLoading(false);
        });
    }
  };

  isTripCreator = () => {
    let isTripCreator = this.context.verifyAuth(this.state.trip[0].user_id);
    return isTripCreator;
  };

  toggleEditTrip = () => {
    this.setState({ toggleEditTrip: false });
  };

  handleDeleteStop = (stop_id) => {
    TripApiService.deleteStop(stop_id)
      .then(() => {
        this.setState({
          stops: this.state.stops.filter((stop) => stop_id !== stop.id),
        });
      })
      .catch((res) => {
        this.setState({ error: res.error });
      });
  };

  render() {
    const trip = this.state.trip[0];
    const stops = this.state.stops.map((stop, index) => {
      if (stop.id === this.state.stopEditingID) {
        return (
          <EditStopForm
            key={stop.id}
            handleSelect={this.handleSelect}
            clearSelections={this.clearSelections}
            selections={this.state.selections}
            handleSubmitEditStop={this.handleSubmitEditStop}
            toggleEditStop={this.toggleEditStop}
            isTripCreator={this.isTripCreator}
            stop={stop}
            index={index}
            stops={this.state.stops}
            error={this.state.error}
          />
        );
      }
      return (
        <RenderStop
          key={stop.id}
          stop={stop}
          index={index}
          toggleEditStop={this.toggleEditStop}
          isTripCreator={this.isTripCreator}
          handleDeleteStop={this.handleDeleteStop}
          stateStops={this.state.stops}
        />
      );
    });
    return (
      <section className="trip-view">
        <div className="trip">
          <div className="tripView-upperSection">
            <div className="trip-info">
              {this.state.toggleEditTrip ? (
                <RenderEditTrip
                  trip={trip}
                  toggleEditTrip={this.toggleEditTrip}
                  handleSubmitEditedTrip={this.handleSubmitEditedTrip}
                />
              ) : (
                <>
                  {!this.context.loading ? (
                    <span className="rating-container">
                      {!this.state.userHasRated ? (
                        <>
                          <Modal
                            isOpen={this.state.isModalOpen}
                            style={Helpers.customModalStyles}
                            contentLabel="Log in!"
                          >
                            <button
                              className="myButton"
                              onClick={() => this.toggleModal(false)}
                            >
                              Close
                            </button>
                            <h3>
                              <center>You must log in to heart a trip!</center>
                            </h3>
                          </Modal>
                          <button
                            className="like-btn"
                            onClick={() => {
                              if (!this.context.returnUserID())
                                this.toggleModal(true);
                              this.handleRating();
                            }}
                          >
                            <img
                              alt="unliked heart"
                              className="empty-heart heart"
                              src={images.EmptyHeart}
                            ></img>
                          </button>
                          <span className="trip-rating-digits">
                            {trip.rating}
                          </span>
                        </>
                      ) : (
                        <>
                          <button className="like-btn">
                            <img
                              alt="liked heart"
                              className="filled-heart heart"
                              src={images.FilledHeart}
                            ></img>
                          </button>
                          <span
                            className="trip-rating-digits"
                            style={{ verticalAlign: 'center' }}
                          >
                            {trip.rating}
                          </span>
                        </>
                      )}
                    </span>
                  ) : null}
                  <h2 className="trip-name">{trip.destination}</h2>
                  <p>{trip.short_description}</p>
                  <p>
                    <span>Activities:</span> {trip.activities} <br />
                    <span>Days:</span> {trip.days}
                  </p>
                </>
              )}
            </div>
            <div id="Map">
              <MapContainer trip={this.state.trip[0]} />
            </div>
          </div>
          <TripViewNav
            handleDeleteTrip={this.handleDeleteTrip}
            handleEditTrip={this.handleEditTrip}
            stops={this.state.stops}
            user_id={this.state.trip[0].user_id}
          />
          <div className="belowMap">
            {stops.length ? (
              stops
            ) : (
              <h4>
                This user hasn't added any stops yet! If this is your trip, add
                some by clicking the Add Stop button!
              </h4>
            )}
            {this.state.toggleAddStop && (
              <AddStopForm
                handleSubmitStop={this.handleSubmitStop}
                handleSelect={this.handleSelect}
                clearSelections={this.clearSelections}
                selections={this.state.selections}
                error={this.state.error}
                toggleAddStop={this.toggleAddStop}
              />
            )}
            {!this.state.toggleAddStop && this.isTripCreator() && (
              <div className="addStopButton">
                <br />
                <div
                  style={{ marginBottom: '20px' }}
                  className="myButton addStopBtn"
                  onClick={() => {
                    this.setState({ toggleAddStop: !this.state.toggleAddStop });
                  }}
                >
                  Add a Stop!
                </div>
              </div>
            )}
          </div>
        </div>
      </section>
    );
  }
}
