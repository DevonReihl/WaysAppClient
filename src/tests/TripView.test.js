import React from 'react';
import TripView from '../components/TripView/TripView';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import TripContext from '../contexts/TripContext';

describe(`TripView component`, () => {
  it('Renders without crashing', () => {
    shallow(
      <TripContext.Provider>
        <div id="root" name="root">
          <TripView match={{ params: { trips_id: 1 } }} />
        </div>
      </TripContext.Provider>
    );
  });
  it('Matches snapshot', () => {
    const wrapper = shallow(
      <TripContext.Provider>
        <TripView match={{ params: { trips_id: 1 } }} />
      </TripContext.Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
