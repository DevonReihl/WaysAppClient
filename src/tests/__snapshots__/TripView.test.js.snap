// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TripView component Matches snapshot 1`] = `
<Trip
  match={
    Object {
      "params": Object {
        "trips_id": 1,
      },
    }
  }
/>
`;
